datafusion/expr/src/logical_plan/extension.rs

```diff
@@ -54,6 +57,22 @@ pub trait UserDefinedLogicalNode: fmt::Debug + Send + Sync {
     /// Return the output schema of this logical plan node.
     fn schema(&self) -> &DFSchemaRef;
 
+    /// Return the list of invariants.
+    ///
+    /// Implementing this function enables the user to define the
+    /// invariants for a given logical plan extension.
+    fn invariants(&self) -> Vec<Invariant> {
+        vec![]
+    }
```

Comment:
This is the new API which can be defined on either `UserDefinedLogicalNode` or `UserDefinedLogicalNodeCore`.